packages/babel-plugin-transform-class-static-block/src/index.ts

```diff
@@ -53,34 +89,81 @@ export default declare(({ types: t, template, assertVersion }) => {
-          // it's common.
-          if (blockBody.length === 1 && t.isExpressionStatement(blockBody[0])) {
-            replacement = t.inheritsComments(
-              blockBody[0].expression,
-              blockBody[0],
-            );
+            if (pendingStaticBlocks.length > 0) {
+              // push static blocks right before the initializer of the next
+              // static field in the class.
+
+              prependToInitializer(
+                path.node,
+                blocksToExpressions(pendingStaticBlocks),
+              );
+              pendingStaticBlocks.length = 0;
+            }
+          }
+        }
+
+        if (pendingStaticBlocks.length > 0) {
+          // if there are static blocks not followed by a static field where
+          // we can inject them, wrap them in a function and push it in the
+          // last static field in the class (or create a new one, if needed).
+          // After the class body runs, call the function to run the remaining
+          // static blocks bodies.
+
+          const tmp = scope.generateDeclaredUidIdentifier("staticBlock");
+          const init = template.expression.ast`${tmp} = () =>
+            ${
+              pendingStaticBlocks.length > 1 ||
+              (pendingStaticBlocks[0].body.length === 1 &&
+                t.isExpressionStatement(pendingStaticBlocks[0].body[0]))
+                ? maybeSequenceExpression(
+                    blocksToExpressions(pendingStaticBlocks),
+                  )
+                : t.blockStatement(pendingStaticBlocks[0].body)
+            }
+          `;
+
+          if (lastStaticProp) {
+            prependToInitializer(lastStaticProp.node, [init]);
           } else {
-            replacement = template.expression.ast`(() => { ${blockBody} })()`;
+            // If there are no static fields at all, it's safe to inject a
+            // new private properties before running the static blocks because
+            // there is no code that could have already made the class
+            // non-extensible.
+
+            const privateNames = new Set<string>();
+            for (const path of body) {
+              if (path.isPrivate()) {
+                privateNames.add(path.get("key.id").node.name);
+              }
+            }
+            const staticBlockPrivateId = generateUid(scope, privateNames);
+            const staticBlockRef = t.privateName(
+              t.identifier(staticBlockPrivateId),
+            );
+            classBody.pushContainer("body", [
+              t.classPrivateProperty(staticBlockRef, init, [], true),
+            ]);
           }
 
-          path.replaceWith(
-            t.classPrivateProperty(
-              staticBlockRef,
-              replacement,
-              [],
-              /* static */ true,
-            ),
+          classBody.parentPath.insertAfter(
+            t.expressionStatement(t.callExpression(t.cloneNode(tmp), [])),
```

Comment:
If the class is anonymous and the `staticBlock` closure is appended, should we also run the `namedEvaluationVisitor` and prepend the `setFunctionName` helper call to the static block statements?

Reply:
BTW, if the class is decorated and there are extra class initializers. Does the `staticBlock` closure now run before the class initializers?

Reply:
> BTW, if the class is decorated and there are extra class initializers. Does the staticBlock closure now run before the class initializers?

The decorators plugin will run before the static blocks plugin (because it visits Class, while this one visits ClassBody) and will inject a static block to run the initializers. Then, this plugin will transform the static block.